### PR TITLE
PLT-3383: image proxy support

### PR DESCRIPTION
**Pull request by ccbrown**
## Summary

This PR adds three new configuration keys: `ImageProxyType`, `ImageProxyURL`, `ImageProxyOptions`.

When these keys are configured, posts served to the client will have their markdown modified such that all images are loaded through a proxy.

For example, if `ImageProxyType` is set to "atmos/camo", this image URL: 

`https://i.redd.it/6cohydckk6501.jpg`

Will be replaced with something like this: 

`https://camo.mysite.com/4cec58f3fc7c74907cbe28adc076ef9650d67325/68747470733a2f2f692e726564642e69742f36636f687964636b6b363530312e6a7067`

(See [atmos/camo](https://github.com/atmos/camo).)

As another example, if `ImageProxyType` is set to "willnorris/imageproxy", that URL will be replaced with something like this:

`https://imageproxy.mysite.com/x500,s5xQmRR3GQa13jFHdJJb01fqATSoZkbBvMXG7Vs4jFGU=/https://i.redd.it/6cohydckk6501.jpg`

(See [willnorris/imageproxy](https://github.com/willnorris/imageproxy).)

## Why?

* Virtually every security questionnaire we get has a checkbox for "no mixed http/https content" that we haven't been able to check off. With image proxies, users make only secure requests. The proxy makes the insecure requests instead.

* By allowing image requests to go straight to third-party servers, we allow things like tracking pixels. Users can put invisible images in their posts that logs time and location data for every user that views the post. An image proxy protects user privacy by eliminating this direct interaction with third-party servers.

* Many proxy servers also provide resizing-as-a-service. If you use willnorris/imageproxy, you can configure `ImageProxyOptions` to resize images if they're more than x pixels tall. So users can save bandwidth by downloading only the resolution they need (This is especially effective with additional client-side support.).

* Proxy servers provide a layer of caching, and can be made faster and more reliable than third-party sites. This caching also helps preserve posts by protecting them from dead images.

## PM Review

There are the three configuration keys I mentioned, and they'll need documentation.

This probably deserves a user interface in the system console in a followup PR.

There is a slight user-facing impact: When you edit posts, image URLs will not be the same as they were when you entered them. For example, after posting `` as a message:

screen shot 2017-12-21 at 4 18 20 am

The message can be edited as usual and if any image URLs were added or changed, they will be rewritten in the same way.

Future API additions and client-side changes could eliminate that impact on editing, but we really wanted the proxying to also work for all existing clients, so this is the only way to do it.

## Dev Review

There are over 6200 lines added to the repo. The good news is that you don't have to review most of these.

The majority of these lines are in the new `utils/markdown` package. This package implements a subset of CommonMark compliant Markdown parsing. This is the best way to accurately locate and rewrite image URLs. You really don't want to try to do that with regex.

This package has hundreds of tests copied straight from the CommonMark spec, which were written specifically to demonstrate all the edge cases they could. As a result, the package has 100% code coverage. ([Codecov](https://codecov.io/gh/mattermost/mattermost-server/pull/7991))

It'll also be useful in closing a plethora of mention / notification issues. I'll put a PR in shortly after this with a new and improved `app.GetExplicitMentions`.

Outside of that package, there are two places of interest:

* model/post.go has a function which uses the Markdown parser to rewrite image URLs given a `func(string)string` transformation.
* app/post.go has functions which wrap image URLs for outgoing posts and unwrap image URLs for incoming posts (So even after editing a post with a proxied image URL, the database post still contains the original, unproxied image URL.).

Aside from those things, there are just one-liners scattered around the API and WebSocket events to wrap / unwrap image URLs.

In the future, if we iterate on this with client support, we'll probably want an API parameter to disable the rewriting so it can be done natively instead.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-3383

#### Checklist
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: model/post.go [hunk 1]
```diff
@@ -390,3 +401,76 @@ func (o *Post) GenerateActionIds() {
...
+	}
+
+	var ranges []markdown.Range
+
+	markdown.Inspect(message, func(blockOrInline interface{}) bool {
+		switch v := blockOrInline.(type) {
+		case *markdown.ReferenceImage:
+			ranges = append(ranges, v.ReferenceDefinition.RawDestination)
+		case *markdown.InlineImage:
+			ranges = append(ranges, v.RawDestination)
```
Comment (ccbrown) [inline model/post.go]: No other Markdown library tracks enough information to actually be able to modify the source without making unnecessary changes to whitespace or syntax. You can inspect the AST with them... but you can't reassemble the original source or find out what source line / column an image URL was defined in.

I really wanted the output of this function to look _exactly_ like the input in every way possible, so I went through great care to make sure you can map block metadata back to source code ranges. It's kinda like the Go parsing packages, except a lot more straight-forward: `.RawDestination` simply provides the range of the raw, unescaped destination string in the original markdown.

File: app/post.go [hunk 2]
```diff
@@ -823,3 +828,124 @@ func (a *App) DoPostAction(postId string, actionId string, userId string) *model
...
+		if strings.HasPrefix(url, proxyURL) {
+			return url
+		}
+
+		if url[0] == '/' {
+			url = siteURL + url
+		}
+
+		switch proxyType {
+		case "atmos/camo":
```
Comment (MusikPolice) [inline app/post.go]: Since these proxy types are configurable, and there are a finite number to choose from, we might consider expressing their names as constants.

File: model/config.go [hunk 3]
```diff
@@ -213,6 +213,9 @@ type ServiceSettings struct {
 	CloseUnusedDirectMessages                *bool
 	EnablePreviewFeatures                    *bool
 	EnableTutorial                           *bool
+	ImageProxyType                           *string
+	ImageProxyURL                            *string
+	ImageProxyOptions                        *string
```
Comment (MusikPolice) [inline model/config.go]: You might consider adding a check to the `isValid()` method to verify that a valid proxy type has been selected. I'd hate to be configuring this, make a spelling mistake, and have the server start up with no complaints, but my proxy not work as desired.

File: app/post.go [hunk 4]
```diff
@@ -823,3 +828,124 @@ func (a *App) DoPostAction(postId string, actionId string, userId string) *model
...
+				mac := hmac.New(sha256.New, []byte(options[1]))
+				mac.Write([]byte(url))
+				digest := base64.URLEncoding.EncodeToString(mac.Sum(nil))
+				if options[0] == "" {
+					return proxyURL + "s" + digest + "/" + url
+				}
+				return proxyURL + options[0] + ",s" + digest + "/" + url
+			}
+			return proxyURL + options[0] + "/" + url
+		}
```
[the remark this answers is not in the document]
Comment (MusikPolice) [reply app/post.go]: I understand all the URL rewriting code, but what I'm not seeing is how the images get loaded into the proxy in the first place. Is the proxy just configured to watch the `data` directory? Is there no need to notify the proxy that a new image has been uploaded?

[the remark this answers is not in the document]
Comment (ccbrown) [reply app/post.go]: The proxies just fetch images on-demand. So the proxies do nothing when the post is created. Then when a user hits the proxy URL, they request the image from its origin, cache it, and return it to the user.

There's actually zero configuration required for the proxy server (except a key if you want signing). E.g. `docker run willnorris/imageproxy` will get you a fully functional proxy server. :)

File: model/config.go (hunk 3 above)
Comment (ccbrown) [reply model/config.go]: +1. I'll be sure to add that.

File: model/post.go [hunk 5]
```diff
@@ -390,3 +401,76 @@ func (o *Post) GenerateActionIds() {
...
+
+var markdownDestinationEscaper = strings.NewReplacer(
+	`\`, `\\`,
+	`<`, `\<`,
+	`>`, `\>`,
+	`(`, `\(`,
+	`)`, `\)`,
+)
+
+func (o *Post) WithRewrittenImageURLs(f func(string) string) *Post {
```
Comment (crspeller) [inline model/post.go]: `f` Is a bad name for a parameter.

File: api4/post.go [hunk 6]
```diff
@@ -211,7 +211,7 @@ func getPost(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	} else {
 		w.Header().Set(model.HEADER_ETAG_SERVER, post.Etag())
-		w.Write([]byte(post.ToJson()))
+		w.Write([]byte(c.App.PostWithProxyAddedToImageURLs(post).ToJson()))
```
Comment (crspeller) [inline api4/post.go]: Maybe we should have this function and the one for lists that is added everywhere be generic. So something like `DoPostRewriting` and have it call PostWithProxyAddedToImageURLs. So if we end up doing more rewriting it's not a pain to add everywhere.

**Comment by crspeller** [comment]
@ccbrown Couple questions:
- Did you write the utils/markdown package yourself?
- Is there any way we can extend this so it's automatic by default? ie. No setup of external services?

File: model/post.go (hunk 5 above)
Comment (ccbrown) [reply model/post.go]: If you think it's unclear, I'm open to renaming or documenting things for clarification, but `f` is at least idiomatic. Just did a search through the standard library and the vast majority of function arguments are simply named `f` (e.g. see the `strings` package). I think the idea is that argument's purpose should be clear from the function name, and if it's not, the function name (not just the argument name) is probably bad.

**Comment by ccbrown** [comment]
@crspeller 

> Did you write the utils/markdown package yourself?

Yeah.

> Is there any way we can extend this so it's automatic by default? ie. No setup of external services?

Yeah, it would be easy to add in a future PR. But for security and scalability reasons, advanced setups will probably always want dedicated servers for their image proxy.

**Comment by crspeller** [comment]
Did you write it yourself because our markdown flavor is too different? Why not use a library? Quick google turned up: https://github.com/russross/blackfriday

File: api4/post.go (hunk 6 above)
Comment (ccbrown) [reply api4/post.go]: I wanted the APIs to explicitly rewrite the image URLs, to facilitate the API parameter we talked about before that would allow disable server-side rewriting and allow client-side URL rewriting.

**Comment by ccbrown** [comment]
> Did you write it yourself because our markdown flavor is too different? Why not use a library? Quick google turned up: https://github.com/russross/blackfriday

Blackfriday can't be used to rewrite Markdown like this. It doesn't track formatting or any source information as it parses. The nodes in the AST that my package generates retains that information, so you can correlate each node to its source markdown.

**Comment by crspeller** [comment]
Ok. I guess nothing like this exists for golang then. Maybe if we keep developing this we can split it out for others to use.

File: api4/post.go (hunk 6 above)
Comment (crspeller) [reply api4/post.go]: Alright. I guess we don't need to solve this problem now.

File: model/post.go (hunk 5 above)
Comment (crspeller) [reply model/post.go]: `func (o *Post) WithRewrittenImageURLs(f func(string) string) *Post`  I don't think it's clear what `f` is for just looking at the signature. I would rename or add a doc comment.

I also hate 1 letter variables anywhere and try to avoid them.

Comment (ccbrown) [reply model/post.go]: Sure. I'll at least add documentation and try to think of other names. 👍

**Comment by jasonblais** [comment]
@ccbrown Once the image proxy is merged, are we able to use it to fix some of our scrollpop issue with link and image previews?

Given all images are loaded through a proxy, not sure if we can get the image/link preview dimensions, add a placeholder and load the image to that placeholder, eliminating scrollpop.

**Comment by ccbrown** [comment]
Nah this doesn't really solve that. The server still doesn't have a way to get the dimensions before sending the post to the client.

**Comment by jasonblais** [comment]
Ah, I see.

**Comment by jasonblais** [comment]
Given this now falls into the ABC team, PM ownership has been transferred to me.

Have flagged for review.

File: utils/markdown/markdown.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+func isHex(c rune) bool {
+	return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
+}
+
+func isHexByte(c byte) bool {
+	return isHex(rune(c))
+}
+
+func nextNonWhitespace(markdown string, position int) int {
+	for offset, c := range []byte(markdown[position:]) {
```
Comment (hmhealey) [inline utils/markdown/markdown.go]: Is it safe to be working with plain bytes when this could involve unicode?

Comment (ccbrown) [reply utils/markdown/markdown.go]: Double check me, but I'm pretty sure all of this is safe. I'm using bytes when possible since it's much faster.

In this case...

> A whitespace character is a space (U+0020), tab (U+0009), newline (U+000A), line tabulation (U+000B), form feed (U+000C), or carriage return (U+000D).

All whitespace characters are single byte runes, so the next non-whitespace rune is any rune that doesn't begin with one of those bytes.

Note the distinction between "whitespace" and "unicode whitespace":

> A Unicode whitespace character is any code point in the Unicode Zs general category, or a tab (U+0009), carriage return (U+000D), newline (U+000A), or form feed (U+000C).

So if this were a "nextNonUnicodeWhitespace" function, it would have to be implemented differently. But unicode whitespace doesn't come into play for any of the parsing that this currently does.

**Comment by ccbrown** [comment]
@hmhealey We could add a new "source_message" field to the API that would be used for editing. /shrug

**Comment by ccbrown** [comment]
@jasonblais Any word on this? If the future of this PR is unknown I might want to break part of it out into a separate PR to help resolve some other issues.

**Comment by jasonblais** [comment]
@ccbrown Sorry for the late reply.

1) There are the three configuration keys I mentioned, and they'll need documentation.

Sounds good. I've added `Needs Docs` label and will work on docs for v4.7.

2) This probably deserves a user interface in the system console in a followup PR.

Is this planned for v4.7? And it will go under Files > Storage?

3) There is a slight user-facing impact: When you edit posts, image URLs will not be the same as they were when you entered them. 

I'm worried. This will not be obvious to end users. Quite a few users use embedded Markdown images (although it's hard to know what %).

4) Future API additions and client-side changes could eliminate that impact on editing, but we really wanted the proxying to also work for all existing clients, so this is the only way to do it.

How much work will this be?

**Comment by jasonblais** [comment]
@ccbrown Are only `ImageProxyType` and `ImageProxyUrl` required fields to set up the image proxy?

**Comment by ccbrown** [comment]
2: I think that's probably a PM decision.
4: I think it would be really easy (a few hours' work).
5: Yes, but `ImageProxyOptions` may also be required depending on `ImageProxyType`.
6: Sure.

**Comment by ccbrown** [comment]
@jasonblais I've deployed a willnorris/imageproxy instance to https://image-proxy.mattermost.com.

I've updated the configuration on that spinmint instance like so:

```
"ImageProxyType": "willnorris/imageproxy",
"ImageProxyURL": "https://image-proxy.mattermost.com",
"ImageProxyOptions": "500,fit"
```

So now if you post a message with an image...

```
![demo](https://willnorris.com/2013/12/small-things.jpg)
```

You'll get a proxied image in your post. This will ensure that every image is downloaded via HTTPS. And in this particular example, having `ImageProxyOptions` set to "500,fit" will also resize images for clients to a maximum of 500 pixels in either dimension.

**Comment by jasonblais** [comment]
2: Proposed texts for System Console page, in Files > Storage, below "Enable Amazon S3 Debugging:" setting. Let me know if you have feedback on the texts.

```
**Image Proxy Type**:
Configure an image proxy to load all Markdown images through a proxy. The image proxy prevents users from making insecure image requests to Mattermost, provides caching for increased performance, and automates image adjustments such as resizing. See [documentation](https://about.mattermost.com/default-image-proxy-documentation) to learn more.

**Image Proxy URL**:
URL of your image proxy server.

**Image Proxy Options**: 
Additional options for basic image adjustments such as resizing, cropping, and rotation. Contact your image proxy service provider to learn more about what options are supported.
```

PS: The texts need to go through PM review, but no need to wait for it. I'll make adjustments directly in the webapp repo if needed.

3/4: If it's easy, I'd propose we take the extra time and do it.
5/6: Thanks! When I post the following message on pre-release, it loads within 1-2 seconds. On the spinmint it took 10 seconds (GIF and the 4K wallpaper took the longest). Is that expected?
```
Mattermost/platform build status:  [![Build Status](https://travis-ci.org/mattermost/platform.svg?branch=master)](https://travis-ci.org/mattermost/platform)  

GitHub favicon:  ![Github](https://assets-cdn.github.com/favicon.ico)

GIF Image:  
![gif](http://i.giphy.com/xNrM4cGJ8u3ao.gif)

4K Wallpaper Image (11Mb):  
![4K Image](http://4kwallpaper.xyz/wallpaper/Large-Galaxy-Lightyears-Space-4K-wallpaper.png)

Panorama Image:  
![Pano](http://amardeepphotography.com/wp-content/uploads/2012/11/Untitled_Panorama6small.jpg)

Tall Portrait Image:  
![Portrait](http://www.maniacworld.com/now-this-is-a-tall-building.jpg)

Tall Portrait Image (resized):  
![Portrait](http://www.maniacworld.com/now-this-is-a-tall-building.jpg =x200)

Tall Portrait Image (resized to a square):  
![Portrait](http://www.maniacworld.com/now-this-is-a-tall-building.jpg =200x200)
```
7: New question: Once the image proxy is configured, will all images load through the proxy, or only the ones posted after configuration?

**Comment by ccbrown** [comment]
@jasonblais 

2: LGTM
3/4: I'll go ahead and make the server-side changes required to support it then.
5/6: I'm not surprised. I didn't expect to be testing this with large images and deployed it on the cheapest hardware possible. It has very little bandwidth, compute power, or memory, and doesn't perform any caching. I can turn up some knobs if we want to. Just costs money.
7: All images.

**Comment by jasonblais** [comment]
5/6: I see, so this is solely due to the size/quality of the proxy server?

Otherwise all sounds good to me.

**Comment by ccbrown** [comment]
5/6: Yep.
